src/shape_transform_descriptor.cpp

```diff
@@ -0,0 +1,830 @@
+    {
+        result.insert(result.end(), dim.subdimensions.begin(), dim.subdimensions.end());
+    }
+    return result;
+}
+
+static std::vector<std::size_t> compute_dims(const operation& op,
+                                             const std::vector<std::size_t>& idims)
+{
+    shape s{shape::float_type, idims};
+    return op.compute_shape({s}).lens();
+}
+
+static std::vector<std::size_t> compute_dims(const std::vector<operation>& ops,
+                                             const std::vector<std::size_t>& idims)
+{
+    shape s{shape::float_type, idims};
+    for(const auto& op : ops)
+        s = op.compute_shape({s});
+    return s.lens();
+}
+
+bool shape_transform_descriptor::apply(const std::vector<operation>& ops)
+{
+    std::vector<std::size_t> dims;
+    std::transform(dimensions.begin(),
+                   dimensions.end(),
+                   std::back_inserter(dims),
+                   [](const dimension& d) { return d.len(); });
+    for(const auto& op : ops)
+    {
+        auto v = op.to_value();
+        if(contains({"reshape", "squeeze", "unsqueeze", "flatten"}, op.name()))
+        {
+            dims = compute_dims(op, dims);
+            if(not apply_reshape(dims))
+                return false;
+        }
+        else if(op.name() == "transpose")
+        {
+            dims = compute_dims(op, dims);
+            if(not apply_transpose(v["permutation"].to_vector<std::int64_t>()))
+                return false;
+        }
+        else if(op.name() == "multibroadcast")
+        {
+            dims = compute_dims(op, dims);
+            // cppcheck-suppress knownConditionTrueFalse
+            if(not apply_broadcast(dims))
+                return false;
+        }
+        else if(op.name() == "broadcast")
+        {
+            dims = compute_dims(op, dims);
+            // cppcheck-suppress knownConditionTrueFalse
+            if(not apply_broadcast(dims, v["axis"].to<std::size_t>()))
+                return false;
+        }
+        else if(op.name() != "contiguous")
+        {
+            return false;
+        }
+    }
+    return true;
+}
+bool shape_transform_descriptor::apply_reshape(const std::vector<std::size_t>& rdims)
+{
+    assert(migraphx::elements(rdims) == this->elements());
+    std::vector<dimension> new_dims;
+    auto subs     = get_all_subdimensions(dimensions);
+    std::size_t i = 0;
+    std::size_t r = 0;
+    while(i < subs.size() and r < rdims.size())
+    {
+        const auto& sub = subs[i];
+        auto idim       = sub.len;
+        auto rdim       = rdims[r];
+        if(idim == rdim)
+        {
+            new_dims.push_back({{sub}});
+        }
+        // squeeze
+        else if(rdim > idim)
+        {
+            auto start = subs.begin() + i;
+            auto it = compute_end_dim(start, subs.end(), rdim, std::mem_fn(&dimension::sub::len));
+            if(it == start)
+                return false;
+            auto n = it - start;
+            i += n;
```

Comment:
```suggestion
            auto n = it - start +1;
            i += n;
```

Reply:
I see you are doing `i++` and `r++` at the bottom to account for additional 1. I find it better to it here.

Reply:
But then I would need to increment the other variables in all branches:

```cpp
        if(idim == rdim)
        {
            ...
            i++;
            r++;
        }
        // squeeze
        else if(rdim > idim)
        {
            ...
            auto n = it - start + 1;
            i += n;
            r++;
            ...
        }
        // unsqueeze
        else // if(rdim < idim)
        {
            ...
            auto n = it - start + 1;
            r += n;
            i++;
            ...
        }
```

It seems simpler to increment them at the end together. This will ensure that I dont forget to increment one of them in one of the branches.